[Step3] 3단계 - 자동차 경주

nooblette opened the pull request:
안녕하세요~! 3단계 미션 PR 요청 드립니다!

작업 내용은 다음과 같습니다.
1. 2단계 리뷰 내용 적용
  - 나눗셈 연산에 대한 검증을 `Sign` Enum 클래스로 이동
  - 나눗셈 연산 검증 위치 수정에 따른 테스트 코드 추가 작성
  - `getSignByString()` 메서드의 로직을 `find()`를 사용하도록 수정


2. 3단계 기능 구현
  - 자동차 개수와 입력 횟수, 출력 함수를 파라미터로 전달 받고 랜덤 숫자 목록을 만들어 경기를 시작하는 RaceClient 클래스를 구현하였습니다.
  - 랜덤 숫자 목록을 받고 자동차 이동을 시작하는 Race 클래스를 구현하였습니다.
  - 거리 상태를 담는 Car 클래스를 선언허고,거리 이동에 따른 테스트 코드 `CarTest`를 작성하였습니다.
  - 숫자 목록을 전달받아 여러대의 Car 객체를 이동시키고 결과를 검증하는 테스트 코드 `RaceTest`를 작성하였습니다.

그리고 RaceClient 클래스의 `startRace()` 메서드에 대한 테스트 코드도 작성하고 싶었으나,
경주 결과를 출력하는 로직을 포함하고 있어 어떻게 검증할 수 있을지 방법이 떠오르지 않아 테스트 코드 작성을 못했는데, 
이 부분의 검증 방법이 궁금합니다.

아니면 미션 요구 사항에서 
**모든 로직에 단위 테스트를 구현한다. 단, UI(System.out, System.in) 로직은 제외**
이 이러한 내용을 말하는 걸까요?

감사합니다 :bow:

nooblette committed:
refactor : 문자열 계산기 코드리뷰 내용 적용

nooblette committed:
feat : 자동차 경주 기능 구현

Hyeon9mak commented on src/main/kotlin/racingcar/car/Car.kt:
```diff
@@ -0,0 +1,11 @@
+package racingcar.car
+
+class Car(
+    val distance: Int = 1,
+) {
+    val result: String
+        get() = "-".repeat(distance)
+
+    // 이동 거리를 1 증가한다.
+    fun increment(): Car = Car(distance + 1)
+}
```
함께 간단한 상상을 해보시죠!

"차량 이동 표현을 `-` 말고 `*` 로 바꿔주세요." 라는 요구사항이 들어왔을 때
동료 개발자는 `ResultView` 와 `Car` 중 어떤 클래스 파일을 먼저 열어보게 될까요?

nooblette replied:
차량 이동 표현은 출력과 관련된 부분이니 동료 개발자는 당연스럽게 `ResultView`를 먼저 열어보게 되고, 해당 기호가 없으니 당황할 것 같네요 😅 

다른 사람 혹은 미래의 내가 봤을때 이해하기 편한 코드가 무엇일지와, 클래스별 역할을 고민해보면 `ResultView`에 두는게 훨씬 적절해 보입니다 👍  고민할 수 있도록 의견 주셔서 감사해요!

Hyeon9mak commented on src/main/kotlin/racingcar/car/Cars.kt:
```diff
@@ -0,0 +1,9 @@
+package racingcar.car
+
+class Cars(
+    private val carCount: Int,
+    val list: List<Car> = List(carCount) { Car() },
+) {
+    // 추가 생성자 구현
+    constructor(list: List<Car>) : this(carCount = list.size, list = list)
+}
```
개발 편의를 위해 다양한 생성자를 만들어주셨군요 👍

nooblette replied:
네 맞습니다 정확히 알아주시다니...! 감사합니다 😄

Hyeon9mak commented on src/main/kotlin/racingcar/client/RaceClient.kt:
```diff
@@ -0,0 +1,26 @@
+    val tryTime: Int,
+    val print: (Cars) -> Unit,
+) {
+    fun startRace() {
+        // 입력받은 자동차 대수로 race 객체 생성
+        var race = Race(cars = Cars(carCount))
+
+        // 시도 횟수를 입력받아 경기 시작
+        println("실행 결과")
+        for (i in 1..tryTime) {
+            // 결과 출력
+            print(race.cars)
+
+            // 경기 시작
+            race = Race(cars = race.run(RandomNumbers(carCount)))
+        }
+    }
```
콘솔을 통해 출력할지, web 화면을 통해 이미지를 이동시킬지 등의 결정은 모두 view 의 책임 아닐까요? 😉

nooblette replied:
출력은 모두 `ResultView`의 책임이고 `RaceClient`는 출력 방식에 의존하면 안된다고 생각해서 `inputView`와 `resultView`를 의존관계 주입받도록 수정했습니다

덕분에 만약 콘솔 출력을 Web 출력으로 바꾸도록 요구사항이 발생해도 `resultView` 필드에 주입할 구현체만 바꿔주면 되니 훨씬 유지보수하기도 편할 것 같아요 👍

Hyeon9mak commented on src/main/kotlin/racingcar/race/Race.kt:
```diff
@@ -0,0 +1,27 @@
+        return Cars(cars.list.mapIndexed { index, car -> move(moveConditions[index], car) })
+    }
+
+    private fun move(
+        isMove: Boolean,
+        car: Car,
+    ): Car =
+        if (isMove) {
+            car.increment()
+        } else {
+            car
+        }
```
이미 충분히 훌륭한 코드인데, 또 다른 생각을 던져보고 싶어요!

> 자동차가 전진하는 조건은 0에서 9 사이에서 무작위 값을 구한 후 무작위 값이 4 이상일 경우이다.

동료 개발자는 `Race` 와 `Car` 중 어떤 클래스를 먼저 열어보려고 할까요?
요런 의사 흐름을 고민해보면 코드를 어디에 배치시킬지 결정하기 편해질거 같네요 😉 👍 

또한.. 테스트 코드 작성도 훨씬 쉬워질 수 있겠어요 👍

nooblette replied:
이부분도 자동차 이동과 상태에 관한 부분이니 `Car` 클래스를 먼저 열어보고 없다면 당황할 것 같네요 😓 
당장 기능 구현만 신경쓰다보니, 동료 개발자가 보기에 편한 코드는 무엇일지에 대한 고민을 놓쳤어요

실무에서도 가끔 놓치곤하는데 좋은 인사이트 주셔서 감사합니다 👍 

의견주신대로 테스트 코드도 더욱 안전하게 작성할 수 있게 됐어요!
```
class CarTest {
    @Test
    fun `숫자가 4 이상이면 자동차가 한 칸 전진한다`() {
        assertThat(Car(DEFAULT_DISTANCE).move(4).distance)
            .isEqualTo(DEFAULT_DISTANCE + 1)
    }

    @Test
    fun `숫자가 4 미만이면 전진하지 않는다`() {
        assertThat(Car(DEFAULT_DISTANCE).move(3).distance)
            .isEqualTo(DEFAULT_DISTANCE)
    }
}
```

Hyeon9mak commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,15 @@
+package racingcar.view
+
+class InputView {
+    fun inputCarCount(): Int {
+        // 자동차 대수 input 받기
+        println("자동차 대수는 몇 대인가요?")
+        return readlnOrNull()?.toInt() ?: 0
+    }
+
+    fun inputTryTime(): Int {
+        // 시도 횟수 input 받기
+        println("시도할 횟수는 몇 회인가요?")
+        return readlnOrNull()?.toInt() ?: 0
+    }
```
별도로 상태가 없는 클래스네요. object 를 활용해볼 수도 있겠는데요? 😉

nooblette replied:
오.. 코틀린은 `object` 라는 키워드로 편리하게 싱글톤 객체를 생성할 수 있군요 🤔 
코틀린의 편리함에 다시 또 놀라고 갑니다..👍  좋은 내용 공유주셔서 감사합니다 😄

Hyeon9mak commented on src/test/kotlin/calculator/StringCalculatorTest.kt:
```diff
@@ -59,4 +59,10 @@ class StringCalculatorTest {
             .isEqualTo(1.5)
     }
+
+    @Test
+    fun `0으로 나눗셈 연산을 시도하는 경우 예외를 던진다`() {
+        assertThatThrownBy { (Sign.DIVIDE.operation(2.0, 0.0)) }
+            .isInstanceOf(IllegalArgumentException::class.java)
+    }
```
😉 👍

Hyeon9mak reviewed (changes requested):
안녕하세요 민혁님! 3단계도 빠르게 잘 진행해주셨네요 👍 

저희가 테스트 코드를 작성하는 이유는 무엇일까요?
결국 "직접 프로그램을 실행시켜보기 전, 불안한 요소에 대해 눈으로 확인해보고 싶다." 는 욕망을 해소하기 위함인거 같아요.
`startRace()` 를 통해서 민혁님께서 확인하실 수 있는 건 어떠한 내용일까요?
'결과 출력이 잘 되는가?' 를 확인하고 싶으셨다면, '꼭 확인할 필요가 있는 복잡하고 중요한 로직인가?' 라고 생각해보셔도 좋을거 같아요.

뭐든 직접 눈으로 확인할 수 있도록 테스트 코드를 작성하는 마음가짐은 중요하지만,
불필요한 테스트를 줄일수록 남아있는 테스트들의 중요성이 더욱 부각되는 효과도 생기니까요 😉 👍 

궁금하신 점 얼마든지 질문 남겨주세요!

nooblette committed:
refactor : 차량 이동을 나타내는 출력 기호를 ResultView로 이동, 출력 기호 상수 선언

nooblette committed:
refactor : RaceClient 의존관계 수정, 콘솔 출력 책임을 ResultView로 이동
- RaceClient 클래스는 inputView와 resultView 타입 객체만 의존관계 주입받으면 된다.
- 만약 콘솔 출력을 Web 출력으로 바꾸더라도 resulView 필드에 주입할 구현체만 바꿔주면 된다.

nooblette committed:
refactor : 값에 따른 자동차 전진 로직을 Car 클래스로 이동 & 추가 테스트 작성

nooblette committed:
refactor : InputView, ResultView 클래스를 object(싱글톤)으로 선언

nooblette committed:
refactor : 시도 횟수 만큼 경기를 반복하는 RaceService 클래스 구현 및 테스트 코드 작성

nooblette commented:
> 저희가 테스트 코드를 작성하는 이유는 무엇일까요? 결국 "직접 프로그램을 실행시켜보기 전, 불안한 요소에 대해 눈으로 확인해보고 싶다." 는 욕망을 해소하기 위함인거 같아요. `startRace()` 를 통해서 민혁님께서 확인하실 수 있는 건 어떠한 내용일까요? '결과 출력이 잘 되는가?' 를 확인하고 싶으셨다면, '꼭 확인할 필요가 있는 복잡하고 중요한 로직인가?' 라고 생각해보셔도 좋을거 같아요.
> 
> 뭐든 직접 눈으로 확인할 수 있도록 테스트 코드를 작성하는 마음가짐은 중요하지만, 불필요한 테스트를 줄일수록 남아있는 테스트들의 중요성이 더욱 부각되는 효과도 생기니까요 😉 👍


제가 무엇을 테스트 코드로 검증하고 싶었는지 다시 고민해보니, 반복 횟수만큼 자동차 경주 반복이 잘 이뤄졌는지 _**최종 경기 결과를 테스트 코드로 검증**_ 하고 싶었던 것 같아요 🤔

이런 고민을 하다보니 시도 횟수와 난수 목록을 파라미터로 전달받아 경기를 반복하는 중간 계층인 `RaceService`를 두게 되었고, 자연스레 테스트 코드는 경기 결과에 대한 검증만 진행하고 출력에 대한 테스트는 신경을 안쓸 수 있게 되었어요!

내가 어떤 내용을 테스트 코드로 검증하고 싶은지를 확실하게 짚는 것이 테스트 코드 작성에서 중요한 부분인것 같네요!
덧붙여서 **불필요한 테스트를 줄일수록 남아있는 테스트들의 중요성이 더욱 부각된다**는 좋은 내용도 공유주셔서 감사합니다 😄

Hyeon9mak commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -0,0 +1,17 @@
+
+import racingcar.car.Cars
+
+object ResultView {
+    private const val PRINT_EXPRESSION = "-"
```
😉 👍

Hyeon9mak commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -0,0 +1,17 @@
+        println("실행 결과")
+    }
+
+    fun printResult(cars: Cars) {
+        // 자동차들의 이동 거리 출력
+        cars.list.forEach { car -> println(PRINT_EXPRESSION.repeat(car.distance)) }
+        println()
+    }
```
repeat 활용 👍

Hyeon9mak commented on src/main/kotlin/racingcar/client/RaceClient.kt:
```diff
@@ -0,0 +1,33 @@
+import racingcar.view.InputView
+import racingcar.view.ResultView
+
+class RaceClient(
+    private val inputView: InputView,
+    private val resultView: ResultView,
+) {
```
이제는 object 가 되었으므로, 별도로 주입받을 필요 없이 바로 상수처럼 호출해서 사용할 수도 있겠네요 😉

Hyeon9mak commented on src/main/kotlin/racingcar/client/RaceClient.kt:
```diff
@@ -0,0 +1,33 @@
+    private val resultView: ResultView,
+) {
+    fun startRace() {
+        // 자동차 대수 입력 받기
+        val carCount = inputView.inputCarCount()
+
+        // 시도 횟수 입력 받기
+        val tryTime = inputView.inputTryTime()
+
+        // 입력받은 carCount만큼 자동차 목록을 갖는 race 객체 생성
```
> 덕분에 만약 콘솔 출력을 Web 출력으로 바꾸도록 요구사항이 발생해도 resultView 필드에 주입할 구현체만 바꿔주면 되니 훨씬 유지보수하기도 편할 것 같아요 👍

💯 👍

Hyeon9mak commented on src/test/kotlin/racingcar/car/CarTest.kt:
```diff
@@ -0,0 +1,22 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class CarTest {
+    @Test
+    fun `숫자가 4 이상이면 자동차가 한 칸 전진한다`() {
+        assertThat(Car(DEFAULT_DISTANCE).move(4).distance)
+            .isEqualTo(DEFAULT_DISTANCE + 1)
+    }
+
+    @Test
+    fun `숫자가 4 미만이면 전진하지 않는다`() {
+        assertThat(Car(DEFAULT_DISTANCE).move(3).distance)
+            .isEqualTo(DEFAULT_DISTANCE)
```
😉 👍

Hyeon9mak commented on src/test/kotlin/racingcar/service/RaceServiceTest.kt:
```diff
@@ -0,0 +1,39 @@
+import racingcar.race.Race
+import racingcar.randomnumber.RandomNumbers
+
+class RaceServiceTest {
+    @Test
+    fun `시도 횟수만큼 자동차 경주를 반복하고 결과를 확인한다`() {
+        val carCount = 3
+        val tryTime = 5
+        val raceService = RaceService(Race(Cars(carCount = carCount)))
+
+        val randomNumbers =
+            List(tryTime) {
+                RandomNumbers(
+                    listOf(
+                        // 1번 자동차만 한칸씩 이동한다.
+                        MOVE_CONDITION,
+                        MOVE_CONDITION - 1,
+                        MOVE_CONDITION - 1,
+                    ),
+                )
+            }
+
+        val raceResult = raceService.execute(tryTime = tryTime, onResult = { }, randomNumbers = randomNumbers)
+
+        // 경기 결과 검증
+        assertThat(raceResult.cars.list)
+            .containsExactly(Car(distance = tryTime), Car(), Car())
+    }
```
"모든 구현 로직이 유기적으로 잘 연결되어있어?" 확인하는 테스트가 되었군요~

Hyeon9mak reviewed (approved):
안녕하세요 민혁님! 피드백 잘 반영해주셨습니다.
4단계로 넘어가도 좋을거 같습니다!
다음 단계에서 뵐게요~
